### DISCUSSION TITLE
Added support for m5stack 4 channel relay and to switch between; Off, Heating or Cooling

### DISCUSSION THREAD

**Discussion by Dennis3321** [raomin/ESPAltherma#296] [Show and tell]
My Altherma device can also cool, and I wanted to control the smart grid. So I needed a total of 4 relays. So I modified the code to include support for the m5stack 4 channel relay.

![image](https://github.com/raomin/ESPAltherma/assets/4771745/6bf2f554-7a4b-4dea-b671-c9ea5579663e)

Made changes to be albe to switch between cooling and heating;

![image](https://github.com/raomin/ESPAltherma/assets/4771745/3a847fce-13b1-4cc7-9336-b4c83230b24b)

If anyone's interested let me know. I've tried to fork this, and tried to upload my code. But dispite me putting my config file in git.ingore it still uploaded my config file online. Stupid ass git.

`{
"availability": [
{
"topic": "espaltherma/LWT",
"payload_available": "Online",
"payload_not_available": "Offline"
}
],
"availability_mode": "all",
"unique_id": "espaltherma_thermostat",
"device": {
"identifiers": [
"ESPAltherma"
],
"manufacturer": "ESPAltherma",
"model": "M5StickC PLUS ESP32-PICO",
"name": "ESPAltherma"
},
"icon": "mdi:thermostat",
"name": "EspAltherma Thermostat",
"command_topic": "espaltherma/thermostat/set",
"command_template": "{% if value == 'Off' %} 0 {% elif value == 'Cooling' %} 1 {% elif value == 'Heating' %} 2 {% else %} 0 {% endif %}",
"options": [
"Off",
"Cooling",
"Heating"
],
"state_topic": "espaltherma/thermostat/state",
"value_template": "{% set mapper = { '0':'Off', '1':'Cooling', '2':'Heating'} %} {% set word = mapper[value] %} {{ word }}"
}`

**Comment by Bertie57** [comment]
Hi - I'm interested in this!  I have the same need and have bought same relay module....   Many thanks

**Comment by Dennis3321** [comment]
I've need to place to put it. I've tried to fork to github and put the passwords in the ignore file, but that didn't work.

**Comment by nolantern** [comment]
Hi, I'm interested in this too.
Could you share how exaclty you've done this?

**Comment by Dennis3321** [comment]
Here you go;

[ESPAltherma_with_unit_4relay.zip](https://github.com/raomin/ESPAltherma/files/12506711/ESPAltherma_with_unit_4relay.zip)

**Comment by andrzejwegrzyn** [comment]
Do you have any diagram for connecting the relays to the pump?

**Reply by Dennis3321** [reply]
You can find that in the manual of your heat pump. Smart grid = 4 wires. Thermostat/ cooling is 4;wires.

**Reply by nolantern** [reply]
It seems that the wiring strongly depends on the hardware that is installed. My model has a diagram printed on the back of the PCB cover.

**Comment by Bertie57** [comment]
Many thanks, I had the same question, I saw the heat pump diagram but
wasn't sure which relay connected to each terminal.

Any way, I'm grateful for your work and your share.

Derry

**Comment by Dennis3321** [comment]
Well, that depends on what model of heat pump you have. You should really look it up in your daikin manual for your model. 

Smart Grid-contact [9.8.5] Bedrijfsmodus Smart Grid
1 2
0 0 Vrij bedrijf
0 1 Gedwongen uit
1 0 Aanbevolen aan
1 1 Gedwongen aan

**Comment by andrzejwegrzyn** [comment]
I tried to compile esp with your code, unfortunately I have this error.
![Zrzut ekranu 2024-07-04 144626](https://github.com/raomin/ESPAltherma/assets/133410597/ef62f45f-487d-4fcf-8879-06eff7a0b684)

**Comment by Jeltel** [comment]
I'm running into the same issue as @andrzejwegrzyn. I have reworked the code example into the latest version of espaltherma so the C plus 2 stick would work.

All seems to be in order except for the undifined references error. The includes are in the right place and the calls look good

@Lectere could you help us out? Or did you find a solution @andrzejwegrzyn ?

**Reply by Jeltel** [reply]
Ah, the .cpp file should be in the src folder, next to the main.cpp and setup.h

It works now with the C plus 2 stick

**Comment by Jeltel** [comment]
@Lectere In the code I don't see any `relay.relayWrite(3, false);` so the fourth relais isn't used. I also don't see a reference to the cooling in the code, except for the line: `#define EEPROM_STATE 0  // off = 0 heating = 1 cooling = 2`

So how did you control the cooling relais?

**Comment by Dennis3321** [comment]
To be honest, I don't remember. Been two years now. Have to dig in again myself.
